localstack/services/s3/provider.py

```diff
@@ -1,12 +1,252 @@
+        response = call_moto_with_exception_patching(context, bucket=request.get("Bucket", ""))
+        return GetObjectOutput(**response, AcceptRanges="bytes")
+
+    @handler("PutObject", expand=False)
+    def put_object(
+        self,
+        context: RequestContext,
+        request: PutObjectRequest,
+    ) -> PutObjectOutput:
+        if checksum_algorithm := request.get("ChecksumAlgorithm"):
+            verify_checksum(checksum_algorithm, context.request.data, request)
+        bucket_name = request.get("Bucket", "")
+        response = call_moto_with_exception_patching(context, bucket_name)
+
+        return PutObjectOutput(**response)
+
+    def get_bucket_lifecycle(
+        self, context: RequestContext, bucket: BucketName, expected_bucket_owner: AccountId = None
+    ) -> GetBucketLifecycleOutput:
+        # TODO: deprecated - both methods have the same URI, what it returns depends on what was put (filter or not)
+        # https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketLifecycle.html
+        response = call_moto_with_exception_patching(context, bucket)
+        return GetBucketLifecycleOutput(**response)
+
+    def get_bucket_lifecycle_configuration(
+        self, context: RequestContext, bucket: BucketName, expected_bucket_owner: AccountId = None
+    ) -> GetBucketLifecycleConfigurationOutput:
+        # TODO: see both methods have the same URI, what it returns depends on what was put (filter or not)
+        # https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketLifecycle.html
+        response = call_moto_with_exception_patching(context, bucket)
+        return GetBucketLifecycleConfigurationOutput(**response)
+
+    def put_bucket_lifecycle(
+        self,
+        context: RequestContext,
+        bucket: BucketName,
+        content_md5: ContentMD5 = None,
+        checksum_algorithm: ChecksumAlgorithm = None,
+        lifecycle_configuration: LifecycleConfiguration = None,
+        expected_bucket_owner: AccountId = None,
+    ) -> None:
+        call_moto_with_exception_patching(context, bucket)
+
+    def delete_bucket_lifecycle(
+        self, context: RequestContext, bucket: BucketName, expected_bucket_owner: AccountId = None
+    ) -> None:
+        call_moto_with_exception_patching(context, bucket=bucket)
+
+
+def call_moto_with_exception_patching(
+    context: RequestContext, bucket: BucketName
+) -> ServiceResponse:
+    try:
+        response = call_moto(context)
+    except CommonServiceException as e:
+        ex = _patch_moto_exceptions(e, bucket_name=bucket)
+        raise ex
+    return response
+
+
+def _patch_moto_exceptions(e: CommonServiceException, bucket_name: BucketName):
+    for exception_class in PATCHED_EXCEPTIONS:
+        if exception_class.code == e.code:
+            ex = exception_class(e.message)
+            ex.BucketName = bucket_name
+            return ex
+    return e
```

Comment:
Is this necessary because moto does not set the `BucketName` or because it's swallowed somewhere down the line in the `call_moto` chain?
If moto actually sets the additional fields on the exception, we should fix it in our forwarding framework.